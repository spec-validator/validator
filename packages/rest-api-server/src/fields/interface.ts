/*
import { Field, SpecUnion, isArraySpec, isFieldSpec, isObjectSpec } from './core'
import { undefinedField, arrayField, objectField } from './fields'

export const getFieldForSpec = <DeserializedType> (
  spec: SpecUnion<DeserializedType>,
  allowExtraFields = false
): Field<DeserializedType> => {
  if (isFieldSpec(spec)) {
    return spec
  } else if (isArraySpec(spec)) {
    return arrayField(getFieldForSpec(spec[0])) as unknown as Field<DeserializedType>
  } else if (isObjectSpec(spec)) {
    return objectField(Object.fromEntries(
      Object.entries(spec).map(([key, value]) => [key, getFieldForSpec(value)])
    ), allowExtraFields) as unknown as Field<DeserializedType>
  } else {
    return undefinedField()
  }
}
*/
