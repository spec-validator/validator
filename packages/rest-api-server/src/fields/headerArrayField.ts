import { declareField, withErrorDecoration } from '@spec-validator/validator/core'
import { FieldWithRegExpSupport } from '@spec-validator/validator/fields/segmentField'

export interface HeaderArrayField<T> extends FieldWithRegExpSupport<T[]> {
  readonly itemField: FieldWithRegExpSupport<T>,
  readonly regex: RegExp,
  serialize(input: T[]): string
}

// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Language
const SEPARATOR = ', '

export default declareField('@spec-validator/rest-api/fields/headerArrayField', <T>(
  itemField: FieldWithRegExpSupport<T>
): HeaderArrayField<T> => {

  const fieldWithStringSupport = itemField.getStringField()
  const regex = fieldWithStringSupport.regex.source

  const result = {
    itemField,
    regex: new RegExp(`${regex}(${SEPARATOR}${regex})+`),
    validate: (value: any): T[] => {
      if (typeof value !== 'string') {
        throw 'Not a string'
      }
      return value.split(SEPARATOR).map(
        (it, index) => withErrorDecoration(index, () => fieldWithStringSupport.validate(decodeURI(it)))
      )
    },
    serialize: (deserialized: T[]): string =>
      deserialized.map(
        (it, index) => withErrorDecoration(index, () => encodeURI(fieldWithStringSupport.serialize(it)))
      ).join(SEPARATOR),
  } as HeaderArrayField<T>

  result.getStringField = () => result

  return result
})
