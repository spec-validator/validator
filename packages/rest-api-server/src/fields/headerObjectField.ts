import { Any } from '@spec-validator/utils/util-types'
import { declareField, TypeHint, withErrorDecoration } from '@spec-validator/validator/core'
import {
  FieldWithStringInputSupport,
} from '@spec-validator/validator/fields/segmentField'

export type HeaderSpec<DeserializedType extends Record<string, Any> = Record<string, Any>> = {
  [P in keyof DeserializedType]: FieldWithStringInputSupport<DeserializedType[P]>
}

export interface HeaderObjectField<Spec extends HeaderSpec = HeaderSpec>
  extends FieldWithStringInputSupport<TypeHint<Spec>> {
  readonly objectSpec: Spec,
  readonly separator: string
  readonly regex: RegExp
}

// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cookie
export default declareField('@spec-validator/rest-api/fields/headerObjectField', <Spec extends HeaderSpec>(
  objectSpec: Spec,
  separator='; '
): HeaderObjectField<Spec> => {

  const regex = '.+=.+'

  const result = {
    objectSpec,
    separator,
    regex: new RegExp(`${regex}(${separator}${regex})*`),
    validate: (value: any): TypeHint<Spec> => {
      if (typeof value !== 'string') {
        throw 'Not a string'
      }
      const payload = Object.fromEntries(value.split(separator)
        .map(it =>it.split(/=(.*)/, 2)))

      return Object.fromEntries(Object.entries(objectSpec).map(([key, valueSpec]) => [
        key, withErrorDecoration(key, () => valueSpec.getFieldWithRegExp().validate(payload[key])),
      ])) as TypeHint<Spec>
    },
    serialize: (deserialized: TypeHint<Spec>): string =>
      Object.entries(objectSpec)
        .map(([key, valueSpec]) => [
          key, withErrorDecoration(key, () => valueSpec.getFieldWithRegExp().serialize(deserialized[key])),
        ])
        .map(([key, value]) => `${key}=${value}`)
        .join(separator),
  } as HeaderObjectField<Spec>

  result.getFieldWithRegExp = () => result

  return result
})
