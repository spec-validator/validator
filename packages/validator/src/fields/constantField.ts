import { Primitive } from '@spec-validator/utils/Json'
import { escapeRegex } from '@spec-validator/utils/utils'
import { declareField } from '../core'
import { FieldWithRegExp, FieldWithRegExpSupport } from './segmentField'

export interface ConstantField<Constant extends Primitive> extends FieldWithRegExpSupport<Constant> {
  readonly constant: Constant
}

export default declareField('@spec-validator/validator/fields/constantField', <Constant extends Primitive> (
  constant: Constant
): ConstantField<Constant> => {

  const result = {
    regexp: escapeRegex(constant.toString()),
    constant,
    validate: (value: any): Constant => {
      if (value !== constant) {
        throw 'Constant does not match the requirement'
      }
      return constant
    },
    serialize: () => constant,
  } as unknown as ConstantField<Constant> & FieldWithRegExp<Constant>
  result.getStringField = () => result

  return result
})
