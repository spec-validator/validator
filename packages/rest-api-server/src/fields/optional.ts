import { Any, Optional } from '@spec-validator/utils/util-types'
import { Field, declareField } from '@spec-validator/validator/core'
import { FieldWithStringInputSupport } from '@spec-validator/validator/fields/segmentField'

export interface OptionalHeaderField<T extends Any> extends Field<Optional<T>> {
  readonly innerField: FieldWithStringInputSupport<T>
}

export default declareField('@spec-validator/rest-api/fields/optional', <T extends Any> (
  innerField: FieldWithStringInputSupport<T>
): OptionalHeaderField<T> => {
  const realField = innerField.getFieldWithRegExp()
  return {
    innerField,
    validate: (value: any): Optional<T> => {
      if (value === undefined) {
        return value
      }
      return realField.validate(value)
    },
    serialize: (deserialized: Optional<T>) =>
      deserialized === undefined ? deserialized : realField.serialize(deserialized) as any,
  }
})
