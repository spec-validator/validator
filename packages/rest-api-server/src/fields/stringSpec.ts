import { Any } from '@spec-validator/utils/util-types'
import { FieldWithStringInputSupport } from '@spec-validator/validator/fields/segmentField'

export type StringObjectSpec<DeserializedType extends Record<string, Any> = Record<string, Any>> = {
  [P in keyof DeserializedType]: FieldWithStringInputSupport<DeserializedType[P]>
}

export type StringArraySpec<DeserializedType extends Any[] = Any[]> = StringSpecUnion<DeserializedType[number]>[]

export type StringSpecUnion<DeserializedType> =
  StringObjectSpec | StringArraySpec | FieldWithStringInputSupport<DeserializedType> | undefined
