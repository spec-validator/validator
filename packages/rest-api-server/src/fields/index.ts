import headerArrayField from './headerArrayField'
import headerObjectField from './headerObjectField'

export { headerObjectField, headerArrayField }
