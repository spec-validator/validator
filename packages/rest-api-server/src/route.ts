import { TypeHint } from '@spec-validator/validator'
import { Field, isFieldSpec, ObjectSpec, SpecUnion, OfType } from '@spec-validator/validator/core'
import { $, unionField } from '@spec-validator/validator/fields'
import { ConstantField } from '@spec-validator/validator/fields/constantField'
import { UnionField } from '@spec-validator/validator/fields/unionField'
import { Any, WithoutOptional } from '@spec-validator/utils/util-types'

export type StringMapping = Record<string, Any>

export type HeaderMapping = Record<string, string | string[] | number>

export type DataMapping = StringMapping | Any

export type RequestSpec<
  Method extends string = string,
  PathParams extends StringMapping | unknown = StringMapping | unknown,
  Body extends Any = Any,
  QueryParams extends StringMapping = StringMapping,
  Headers extends HeaderMapping = HeaderMapping,
> = {
  readonly method: ConstantField<Method>,
  readonly pathParams: typeof $ & Field<PathParams>,
  readonly body?: SpecUnion<Body>,
  readonly headers?: ObjectSpec<Headers>,
  readonly queryParams?: ObjectSpec<QueryParams>
}

export type ResponseSpec<
  StatusCode extends number = number,
  Body extends Any = Any,
  Headers extends HeaderMapping = HeaderMapping,
> = {
  readonly statusCode: ConstantField<StatusCode>,
  readonly body?: SpecUnion<Body>,
  readonly headers?: ObjectSpec<Headers>,
}

type ResponseField<Spec extends ResponseSpec=ResponseSpec > = WithoutOptional<Spec>

// TODO: how to extract schema from
type ResponsesSpec<ResponseVariants extends ResponseField[] = ResponseField[]> =
  UnionField<ResponseVariants>

export type Route<
  ReqSpec extends RequestSpec = RequestSpec,
  RespSpec extends ResponsesSpec | ResponseSpec = ResponsesSpec | ResponseSpec
> = {
  readonly request: ReqSpec,
  readonly response: RespSpec,
  readonly handler: (request: WithoutOptional<TypeHint<ReqSpec>>) => Promise<
    WithoutOptional<TypeHint<RespSpec>>
  >
}

export const isResponsesSpec = (spec: ResponsesSpec | ResponseSpec): spec is ResponsesSpec =>
  isFieldSpec(spec) && (spec as unknown as OfType<string>).type !== unionField.type

export const route = <
  ReqSpec extends RequestSpec = RequestSpec,
  RespSpec extends ResponsesSpec | ResponseSpec = ResponsesSpec | ResponseSpec,
  >(spec: {
    request: ReqSpec,
    response: RespSpec
  }): {
    handler: (handler:
      (request: WithoutOptional<TypeHint<ReqSpec>>) => Promise<WithoutOptional<TypeHint<RespSpec>>>
    ) => Route
  } => ({
    handler: (handler): Route => ({
      ...spec,
      handler: async (request) => await handler(request as any),
    }),
  })
