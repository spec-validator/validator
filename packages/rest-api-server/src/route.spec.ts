import { choiceField, numberField, objectField, stringField, unionField } from '@spec-validator/validator/fields'
import constantField from '@spec-validator/validator/fields/constantField'
import $ from '@spec-validator/validator/fields/segmentField'

import { expectType } from '@spec-validator/test-utils/expectType'
import { Route } from './route'

describe('Route', () => {

  it('always contains the fields that are defined', () => {
    const reqSpec = {
      method: constantField('GET'),
      headers: {
        headerKey: stringField(),
      },
      pathParams: $._('pathKey', stringField()),
    }

    const respSpec = {
      statusCode: constantField(201),
      data: stringField(),
    }

    type Decor = Route<typeof reqSpec, typeof respSpec>

    expectType<Decor, {
      request: typeof reqSpec,
      response: typeof respSpec,
      handler: (request: {
        method: 'GET',
        headers: { headerKey: string },
        pathParams: { pathKey: string },
      }) => Promise<{
        statusCode: 201,
        data: string
      }>
        }>(true)
  })

  it('always contains a method in a request and statusCode in a response', () => {
    const reqSpec = {
      method: constantField('GET'),
      pathParams: $,
    }

    const respSpec = {
      statusCode: constantField(201),
    }

    type Decor = Route<typeof reqSpec, typeof respSpec>

    expectType<Decor, {
      request: typeof reqSpec,
      response: typeof respSpec,
      handler: (request: {
        method: 'GET',
      }) => Promise<{
        statusCode: 201,
      }>
        }>(true)
  })

  it('works with a union of responses', () => {
    const reqSpec = {
      method: constantField('GET'),
      pathParams: $,
    }

    const respSpec = unionField({
      statusCode: constantField(201),
      data: numberField(),
    }, {
      statusCode: constantField(202),
      data: choiceField('one', 'two'),
      headers: objectField({
        headerKey: stringField(),
      }),
    })

    type Decor = Route<typeof reqSpec, typeof respSpec>

    expectType<Decor, {
      request: typeof reqSpec,
      response: typeof respSpec,
      handler: (request: {
        method: 'GET',
      }) => Promise<{
        statusCode: 201,
        data: number
      } | {
        statusCode: 202,
        data: 'one' | 'two',
        headers: {
          headerKey: string
        }
      }>}>(true)
  })
})
