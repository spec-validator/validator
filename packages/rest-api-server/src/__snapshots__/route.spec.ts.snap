// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`route function returns a proper route object 1`] = `
Object {
  "headers": Object {
    "headerKey": Object {
      "getStringField": [Function],
      "regex": /\\.\\*/,
      "serialize": [Function],
      "type": "@spec-validator/validator/fields/stringField",
      "validate": [Function],
    },
  },
  "method": Object {
    "constant": "GET",
    "getStringField": [Function],
    "regex": /GET/,
    "serialize": [Function],
    "type": "@spec-validator/validator/fields/constantField",
    "validate": [Function],
  },
  "pathParams": SegmentField {
    "field": Object {
      "regex": /\\.\\*/,
      "serialize": [Function],
      "type": "@spec-validator/validator/fields/stringField",
      "validate": [Function],
    },
    "key": "pathKey",
    "parent": SegmentField {
      "field": undefined,
      "key": "",
      "parent": undefined,
      "type": "@spec-validator/validator/fields/segmentField",
    },
    "type": "@spec-validator/validator/fields/segmentField",
  },
}
`;

exports[`route function returns a proper route object 2`] = `
Object {
  "serialize": [Function],
  "type": "@spec-validator/validator/fields/unionField",
  "validate": [Function],
  "variants": Array [
    Object {
      "body": Object {
        "canBeFloat": false,
        "getStringField": [Function],
        "regex": /\\\\d\\+/,
        "serialize": [Function],
        "signed": false,
        "type": "@spec-validator/validator/fields/numberField",
        "validate": [Function],
      },
      "statusCode": Object {
        "constant": 201,
        "getStringField": [Function],
        "regex": /201/,
        "serialize": [Function],
        "type": "@spec-validator/validator/fields/constantField",
        "validate": [Function],
      },
    },
    Object {
      "body": Object {
        "choices": Array [
          "one",
          "two",
        ],
        "getStringField": [Function],
        "regex": /0\\|1/,
        "serialize": [Function],
        "type": "@spec-validator/validator/fields/choiceField",
        "validate": [Function],
      },
      "headers": Object {
        "headerKey": Object {
          "getStringField": [Function],
          "regex": /\\.\\*/,
          "serialize": [Function],
          "type": "@spec-validator/validator/fields/stringField",
          "validate": [Function],
        },
      },
      "statusCode": Object {
        "constant": 202,
        "getStringField": [Function],
        "regex": /202/,
        "serialize": [Function],
        "type": "@spec-validator/validator/fields/constantField",
        "validate": [Function],
      },
    },
  ],
}
`;
